8337702: Use new ForwardExceptionNode to call StubRoutines::forward_exception_entry()

vnkozlov opened the pull request:
Currently C2 uses `TailCall` node when it generates code to forward exception in C2 runtime stubs.
`StubRoutines::forward_exception_entry()` address is passed as constant and  method pointer is `NULL`: 
[generateOptoStub.cpp#L258](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/generateOptoStub.cpp#L258)

On other hand TailCall mach node uses 2 registers as parameter which is hardcoded in `Matcher`: [matcher.cpp#L828](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/matcher.cpp#L828)
As result we waste two registers to pass constant and NULL.

Also incorrect relocation is used for such call because the address of `forward_exception` stub passed in register in mach node. When it is converted to `Address` for `jmp` instruction the default `external_word_type` relocation is used when `runtime_call_type` should be used. See discussion in PR [JDK-8337396](https://github.com/openjdk/jdk/pull/20412)

I added new ideal node `ForwardExceptionNode` to solve these issues. It is similar to `Rethrow` node (which mach node definition I used as template) but I kept it based on `Return` node similar to `TailCall` node.

Tested tier1-3,stress,xcomp

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337702](https://bugs.openjdk.org/browse/JDK-8337702): Use new ForwardExceptionNode to call StubRoutines::forward_exception_entry() (**Enhancement** - P4)


### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20437/head:pull/20437` \
`$ git checkout pull/20437`

Update a local copy of the PR: \
`$ git checkout pull/20437` \
`$ git pull https://git.openjdk.org/jdk.git pull/20437/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20437`

View PR using the GUI difftool: \
`$ git pr show -t 20437`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20437.diff">https://git.openjdk.org/jdk/pull/20437.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20437#issuecomment-2264628944)

adinn reviewed (commented):
Hi Vladimir.

This looks like a very nice solution. The code changes look ok as far a I can tell. However, I'm not really in a position to approve the patch as I am not very familiar with the details of the matcher and formssel code. Sorry.

vnkozlov commented:
Thank you, @adinn, for review.

RealFYang commented:
FYI: Also performed tier1-3 test on linux-riscv64. Result looks good.

TobiHartmann commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -152,6 +152,17 @@ class RethrowNode : public Node {
+// Pop stack frame and jump to StubRoutines::forward_exception_entry()
+class ForwardExceptionNode : public ReturnNode {
+public:
+  ForwardExceptionNode( Node *cntrl, Node *i_o, Node *memory, Node *frameptr, Node *retadr)
+    : ReturnNode( TypeFunc::Parms, cntrl, i_o, memory, frameptr, retadr ) {
```
```suggestion
  ForwardExceptionNode(Node* cntrl, Node* i_o, Node* memory, Node* frameptr, Node* retadr)
    : ReturnNode(TypeFunc::Parms, cntrl, i_o, memory, frameptr, retadr) {
```

TobiHartmann commented on src/hotspot/cpu/aarch64/aarch64.ad:
```diff
@@ -16184,6 +16184,19 @@ instruct TailjmpInd(iRegPNoSpNoRfp jump_target, iRegP_R0 ex_oop)
 %}
 
+// Forward exception.
+instruct ForwardExceptionjmp()
```
```suggestion
instruct ForwardException()
```

Same for other AD files.

vnkozlov replied:
We can't use the same name for Mach node as for Ideal node:
```
src/hotspot/cpu/x86/x86_64.ad(12596) Syntax Error: :duplicate name ForwardException for instruction
Error Context:  >>>(<<<)
src/hotspot/cpu/x86/x86_64.ad(12597) Syntax Error: :Identifier expected, but found '%{
  match(ForwardEx[...]'.
Error Context:  >>>%<<<{
```

TobiHartmann replied:
Ah, makes sense.

TobiHartmann reviewed (approved):
Nice enhancement. 

I noticed that `TailJump` is in vmStructs.cpp, should `ForwardException` be added as well?

vnkozlov commented:
Thank you, @RealFYang, for testing.

vnkozlov commented:
Hi, @TobiHartmann 

I cleaned `ForwardExceptionNode` constructor using your suggestion and added its type to `vmStruct`.

I kept name of Mach instructions unchanged (with `jump` at the end) because we can't use the same name as Ideal node.  We  do generate `jump' instruction so the name is appropriate in this sense I think.

TobiHartmann reviewed (approved):
Looks good!

ichttt commented on src/hotspot/share/opto/generateOptoStub.cpp:
```diff
@@ -256,12 +256,11 @@ void GraphKit::gen_stub(address C_function,
 
   assert (StubRoutines::forward_exception_entry() != nullptr, "must be generated before");
   Node *exc_target = makecon(TypeRawPtr::make( StubRoutines::forward_exception_entry() ));
```
`exc_target` is no longer used, so this should probably be removed

vnkozlov replied:
removed

TobiHartmann reviewed (approved):
Still good.

vnkozlov commented:
Thank you, Tobias

vnkozlov commented:
/integrate
